cmd/erasure-object.go

```diff
@@ -630,91 +683,145 @@ func (xl xlObjects) putObject(ctx context.Context, bucket string, object string,
+		partsMetadata[index].Size = n
+		partsMetadata[index].ModTime = modTime
 	}
 
-	// Write unique `xl.json` for each disk.
-	if onlineDisks, err = writeUniqueXLMetadata(ctx, onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, writeQuorum); err != nil {
+	// Write unique `xl.meta` for each disk.
+	if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, writeQuorum); err != nil {
 		return ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
 	// Rename the successfully written temporary object to final location.
-	if onlineDisks, err = rename(ctx, onlineDisks, minioMetaTmpBucket, tempObj, bucket, object, true, writeQuorum, nil); err != nil {
+	if onlineDisks, err = renameData(ctx, onlineDisks, minioMetaTmpBucket, tempObj, fi.DataDir, bucket, object, writeQuorum, nil); err != nil {
 		return ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
 	// Whether a disk was initially or becomes offline
 	// during this upload, send it to the MRF list.
 	for i := 0; i < len(onlineDisks); i++ {
 		if onlineDisks[i] == nil || storageDisks[i] == nil {
-			xl.addPartialUpload(bucket, object)
+			er.addPartialUpload(bucket, object)
 			break
 		}
 	}
 
 	// Object info is the same in all disks, so we can pick the first meta
 	// of the first disk
-	xlMeta = partsMetadata[0]
+	fi = partsMetadata[0]
 
 	objInfo = ObjectInfo{
 		IsDir:           false,
 		Bucket:          bucket,
 		Name:            object,
-		Size:            xlMeta.Stat.Size,
-		ModTime:         xlMeta.Stat.ModTime,
-		ETag:            xlMeta.Meta["etag"],
-		ContentType:     xlMeta.Meta["content-type"],
-		ContentEncoding: xlMeta.Meta["content-encoding"],
-		UserDefined:     xlMeta.Meta,
+		Size:            fi.Size,
+		ModTime:         fi.ModTime,
+		ETag:            fi.Metadata["etag"],
+		ContentType:     fi.Metadata["content-type"],
+		ContentEncoding: fi.Metadata["content-encoding"],
+		UserDefined:     fi.Metadata,
 	}
 
 	return objInfo, nil
 }
 
+func (er erasureObjects) deleteObjectDir(ctx context.Context, bucket, object string, writeQuorum int) error {
+	// Rename the current object while requiring write quorum, but also consider
+	// that a non found object in a given disk as a success since it already
+	// confirms that the object doesn't have a part in that disk (already removed)
+	tmpObj := mustGetUUID()
+	disks, err := rename(ctx, er.getDisks(), bucket, object, minioMetaTmpBucket, tmpObj, true, writeQuorum,
+		[]error{errFileNotFound})
+	if err != nil {
+		return toObjectErr(err, bucket, object)
+	}
+
+	g := errgroup.WithNErrs(len(disks))
+
+	for index := range disks {
+		index := index
+		g.Go(func() error {
+			if disks[index] == nil {
+				return errDiskNotFound
+			}
+			// DeleteFile() simply tries to remove a directory
+			// and will succeed only if that directory is empty.
+			err := disks[index].DeleteFile(minioMetaTmpBucket, tmpObj)
+			if err != nil && err != errVolumeNotFound {
```

Comment:
i.e because errVolumeNotFound is returned if its a directory @vadmeste

Reply:
> i.e because errVolumeNotFound is returned if its a directory @vadmeste

I think errVolumeNotFound is only related to buckets (minioMetaTmpBucket in this case). I just took a look on DeleteFile in xl-storage.go